server/service/oauth_test.go

```diff
@@ -0,0 +1,115 @@
+package service
+
+import (
+	"net/http"
+	"testing"
+
+	"github.com/src-d/code-annotation/server/model"
+	"github.com/stretchr/testify/suite"
+)
+
+// roleTestResult is used instead of models.Role to have the "denied" option
+type roleTestResult string
+
+const (
+	denied    roleTestResult = "denied"
+	worker    roleTestResult = "worker"
+	requester roleTestResult = "requester"
+)
+
+const (
+	outsiderGroup  = "NONE"
+	accessGroup    = "org:accessGroup"
+	requesterGroup = "org:requesterGroup"
+)
+
+// testRestrictionChecker implements restrictionChecker interface. The mock
+// check is done comparing the user's login name with the group restriction
+type testRestrictionChecker struct{}
+
+func (checker *testRestrictionChecker) checkAccess(client *http.Client, restriction, login string) error {
+	if login != restriction {
+		return ErrNoAccess
+	}
+
+	return nil
+}
+
+var testChecker testRestrictionChecker
+
+// getTestResult is a helper that performs oauth.setRole
+func getTestResult(oauth *OAuth, user githubUser) roleTestResult {
+	err := oauth.setRole(&testChecker, nil, &user)
+
+	if err != nil {
+		return denied
+	}
+
+	switch user.Role {
+	case model.Requester:
+		return requester
+	case model.Worker:
+		return worker
+	default:
+		return denied
+	}
+}
+
+type OAuthSuite struct {
+	suite.Suite
+}
+
+/*
+TestGetUser tests role assignment based on the group membership. The tested
+logic is the following table:
+
++-------------------+-----------------------------+-------------+---------------+------------------+
+|                   |                             |             |  User in      |  User in         |
+|  RESTRICT_ACCESS  |  RESTRICT_REQUESTER_ACCESS  |  Outsider   |  accessGroup  |  requesterGroup  |
++-------------------+-----------------------------+-------------+---------------+------------------+
+|  -                |  -                          |  Requester  |  N/A          |  N/A             |
+|  -                |  requesterGroup             |  Worker     |  N/A          |  Requester       |
+|  accessGroup      |  -                          |  Denied     |  Requester    |  N/A             |
```

Comment:
Because that column refers to the users members of the group set in in RESTRICT_REQUESTER_ACCESS. If there isn't one set, you can't reason about the users belonging to that group. Hence, N/A.

Does it make sense?

Reply:
I understood from your explanation irl that the table was the key to understand the testcases below; if so, I think it would be useful to have a complete table.
For example, from the second row:
> if the **requester rule is defined,** and the user trying to access is in the **group of people that can access** the app, the assigned role can be **whatever**

And it does not match with the real test case, that expects the **worker** role for that user.

Reply:
I have just pushed e8673cba1b4b16a9c9aaf11bdbb9b3cee7cede0e to add the N/A option as a test result.

By the way, just to clarify any confusion, the second row does not mean:
> if the **requester rule is defined,** and the user trying to access is in the **group of people that can access** the app, the assigned role can be **whatever**

```
+-------------------+-----------------------------+-------------+---------------+------------------+
|                   |                             |             |  User in      |  User in         |
|  RESTRICT_ACCESS  |  RESTRICT_REQUESTER_ACCESS  |  Outsider   |  accessGroup  |  requesterGroup  |
+-------------------+-----------------------------+-------------+---------------+------------------+
|  -                |  requesterGroup             |  Worker     |  N/A          |  Requester       |
```

In this case there isn't a _group of people that can access the app_. There are only users on the `requesterGroup`, and people that are not in it. Members of the `accessGroup`do not hold any special meaning in this test case, because they are not part of the restriction set in `RESTRICT_ACCESS`, and the test is _not applicable_.
But it does not mean _whatever_, of course. As any other user not in the `requesterGroup`, these users fall into the outsider column, that is, `worker` role.

Reply:
thanks!